pkg/agent/flexvolume/controller.go

```diff
@@ -172,11 +188,15 @@ func (c *FlexvolumeController) Detach(detachOpts AttachOptions, _ *struct{} /* v
 
 	namespace := os.Getenv(k8sutil.PodNamespaceEnvVar)
 	crdName := detachOpts.VolumeName
-	logger.Infof("Deleting Volumeattachment CRD %s/%s", namespace, crdName)
-	return c.volumeAttachmentController.Delete(namespace, crdName)
+	volumeAttach, err := c.volumeAttachmentController.Get(namespace, crdName)
+	if len(volumeAttach.Attachments) == 0 {
+		logger.Infof("Deleting VolumeAttachment CRD %s/%s", namespace, crdName)
+		return c.volumeAttachmentController.Delete(namespace, crdName)
```

Comment:
we are only deleting by the name of the CRD, which doesn't capture any version information.  Therefore, someone else could have updated the CRD in between the Get, our local manipulation, then our call to Delete by name.  I know that's a very slim window, but to harden this we should also ensure that the object on the API side is the version that we expect it to be from this client's perspective when we are calling Delete here.

I thought we would have the same problem with the Update call just below in `RemoveAttachmentObject()` but I think we are OK in that case because we are not updating by name, we are updating by object.  Our definition for `VolumeAttachment` has a `ObjectMeta` field, which has a resource version field.  Can you confirm for me that we our CRD `Update()` operations are safe with respect to only updating if the API side object is the version we expect it to be?

Reply:
yes. the update() will fail if we are trying to update something that has already changed.

As for the small window where the crd could change. I am not sure how we can do this since the get-delete operation cannot be made atomic. What we could do is to minimize this window. Which i think is pretty minimal now. I am doing a get, then check the length of the attachment list and deleting it based on the length.

One thing we can do is not to delete the Volumeattach crd. We can just leave it blank. We can delete it once users deletes the volume.

Reply:
Could we simply pass the volume attach object we got from the `Get()` call to the `Delete()` call instead of passing just its name?  Would that be an easy way to take advantage of the version checking that k8s/etcd providers for us?

Reply:
I dont think you can do that. For create() and update() you can because the REST api takes a body. For things like get/delete, you can only do that by name.

Reply:
I think you're right.  Specifying the expected object version for DELETE is not currently supported according to https://github.com/kubernetes/community/blob/master/contributors/devel/api-conventions.md#concurrency-control-and-consistency:
```
resourceVersion may be used as a precondition for other operations (e.g., GET, DELETE) in the future, such as for read-after-write consistency in the presence of caching.
```

Reply:
im not sure how they will do this for delete since for restful API you can only specify a name when using the HTTP verb DELETE.